### PR TITLE
Integrate frontpage html from aptoma

### LINKED ISSUE #841: Render html frontpages from lettera
Some front pages have a pre-rendered html version which we serve from **lettera**. **mosaico** should use this version when available.

### PR DESCRIPTION
**Pull request by ae-mo**
Closes #841.

### CODE REVIEW

File: apps/mosaico/package.json [hunk 1]
```diff
@@ -20,6 +19,6 @@
   "scripts": {
     "start": "node -r dotenv/config -e 'require(\"./output/Main/index\").main()' dotenv_config_path=.env.local",
     "start-dev": "parcel web/index.html -p 8000 -d dist/client --public-url /assets && parcel watch ./static/* -d dist/client",
-    "build": "spago build && parcel build web/index.html -d dist/client --public-url /assets && parcel build ./static/* -d dist/client"
+    "build": "parcel build ./static/* -d dist/client && parcel watch web/index.html -d dist/client --public-url /assets"
```
Comment (ae-mo) [inline apps/mosaico/package.json:22]: Sorry about this, gonna revert

File: apps/mosaico/src/Main.purs [hunk 2]
```diff
@@ -307,9 +311,27 @@ frontpage env { guards: { credentials } } = do
             appendMosaico mosaicoString env.htmlTemplate >>= appendHead (mkWindowVariables windowVars)
   pure $ maybeInvalidateAuth user $ htmlContent $ Response.ok $ StringBody html
 
-mkArticleFeed :: Maybe String -> String -> Array ArticleStub -> String
-mkArticleFeed feedPage feedType feedContent =
-  stringify $ encodeJson { feedPage, feedType, feedContent: encodeStringifyArticleStubs feedContent }
+getFrontPage :: Paper -> String -> Aff ArticleFeed
+getFrontPage paper category = do
```
Comment (kaol) [inline apps/mosaico/src/Main.purs:315]: Remove trailing white space.

File: apps/mosaico/src/Mosaico.purs [hunk 3]
```diff
@@ -294,8 +283,14 @@ fromJSProps jsProps =
             "searchfeed"   -> SearchFeed <$> feedPage
             _              -> Nothing
         feedContent <- do
-          content <- toMaybe feed.feedContent >>= (jsonParser >>> hush) >>= toArray
-          pure $ mapMaybe (hush <<< parseArticleStubWithoutLocalizing) content
+          content <- toMaybe feed.feedContent
+          feedContentType <- toMaybe feed.feedContentType
+          case feedContentType of
```
Comment (kaol) [inline apps/mosaico/src/Mosaico.purs:288]: Remove trailing white space.

File: apps/mosaico/src/Mosaico.purs [hunk 4]
```diff
@@ -449,11 +448,16 @@ render setState state components router onPaywallEvent =
     doSearch query = do
       router.pushState (write {}) $ "/sök?q=" <> query
 
-    frontpage frontpageArticles = mosaicoDefaultLayout $ components.frontpageComponent
-      { frontpageArticles
-      , onArticleClick
-      , onTagClick
-      }
+    frontpage frontpageArticles =
```
Comment (kaol) [inline apps/mosaico/src/Mosaico.purs:451]: Remove trailing white space.

File: apps/mosaico/src/Mosaico/Frontpage.purs [hunk 5]
```diff
@@ -10,29 +10,51 @@ import Data.Newtype (un)
...
 
-render :: Props -> JSX
-render props =
-   DOM.div
+render :: State -> Props -> JSX
+render state props =
+  DOM.div
     { className: "mosaico--article-list"
-    , children: maybe [loadingSpinner] (map renderListArticle) props.frontpageArticles
+    , children: maybe
```
Comment (kaol) [inline apps/mosaico/src/Mosaico/Frontpage.purs:44]: Remove trailing white space.

File: apps/mosaico/src/Mosaico/Frontpage.purs [hunk 6]
```diff
@@ -10,29 +10,51 @@ import Data.Newtype (un)
...
+render :: State -> Props -> JSX
+render state props =
+  DOM.div
     { className: "mosaico--article-list"
-    , children: maybe [loadingSpinner] (map renderListArticle) props.frontpageArticles
+    , children: maybe 
+        [loadingSpinner]
+        (\content -> case content of
+            ArticleList list -> map renderListArticle list
+            Html html        -> [ state.htmlRendererComponent
```
Comment (kaol) [inline apps/mosaico/src/Mosaico/Frontpage.purs:48]: Remove trailing white space.

File: apps/mosaico/src/Mosaico/Frontpage.purs [hunk 7]
```diff
@@ -10,29 +10,51 @@ import Data.Newtype (un)
...
+            Html html        -> [ state.htmlRendererComponent 
+                                    { content: html
+                                    , hooks: Just [ andraLaserHook
+                                                  ]
+                                    }
+                                ]
+        )
+        props.content
     }
+
```
Comment (kaol) [inline apps/mosaico/src/Mosaico/Frontpage.purs:57]: Remove white space.

File: packages/components/src/HtmlRenderer/Models.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
+"use strict";
+
+// html-to-react uses htmlparser to parse input and each
```
Comment (kaol) [inline packages/components/src/HtmlRenderer/Models.js:3]: Remove trailing white space.

File: packages/components/src/HtmlRenderer/Models.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+exports.getNameImpl = (node) => {
+    if (node && node.name)
+      return node.name;
+
+    return null;
+}
+
+exports.getAttribsImpl = (node) => {
```
Comment (kaol) [inline packages/components/src/HtmlRenderer/Models.js:35]: Remove trailing white space from this line and from the rest of exports in this file too.

File: packages/components/src/HtmlRenderer/Models.purs [hunk 10]
```diff
@@ -0,0 +1,639 @@
...
+
+-- | A representation of a hook, can be used to group different
+--   type of hooks in the same collection
+--   See: https://stackoverflow.com/questions/53270182/similar-record-types-in-a-list-array-in-purescript
+newtype HookRep = HookRep (forall a. (forall h. ToGenericHook h => h -> a) -> a)
+
+toHookRep :: forall h. ToGenericHook h => h -> HookRep
+toHookRep h = HookRep \f -> f h
+
+-- | This type of hook is to replace the children of
```
Comment (kaol) [inline packages/components/src/HtmlRenderer/Models.purs:72]: Remove trailing white space.

File: packages/components/src/HtmlRenderer/Models.purs [hunk 11]
```diff
@@ -0,0 +1,639 @@
...
+newtype HookRep = HookRep (forall a. (forall h. ToGenericHook h => h -> a) -> a)
+
+toHookRep :: forall h. ToGenericHook h => h -> HookRep
+toHookRep h = HookRep \f -> f h
+
+-- | This type of hook is to replace the children of 
+--   targeted nodes
+newtype ReplacingHook = ReplacingHook ReplacingHookRecord
+
+type ReplacingHookRecord =
```
Comment (kaol) [inline packages/components/src/HtmlRenderer/Models.purs:76]: Remove trailing white space.

File: packages/components/src/HtmlRenderer/Models.purs [hunk 12]
```diff
@@ -0,0 +1,639 @@
...
+type ReplacingHookRecord = 
+  { shouldProcessNode :: Node -> Boolean
+  , processNode       :: Node -> Array Node -> String -> Effect JSX
+  }
+
+replacingHook :: ReplacingHookRecord -> HookRep
+replacingHook = toHookRep <<< ReplacingHook
+
+instance replacingHookToGenericHook :: ToGenericHook ReplacingHook where
+  toGenericHook (ReplacingHook { shouldProcessNode, processNode }) =
```
Comment (kaol) [inline packages/components/src/HtmlRenderer/Models.purs:85]: Remove trailing white space.

File: packages/components/src/HtmlRenderer/Models.purs [hunk 13]
```diff
@@ -0,0 +1,639 @@
...
+type ModifyingHookRecord =
+  { shouldProcessNode :: Node -> Boolean
+  , processNode       :: Node -> Array Node -> Effect Unit
+  }
+
+modifyingHook :: ModifyingHookRecord -> HookRep
+modifyingHook = toHookRep <<< ModifyingHook
+
+instance modifyingHookToGenericHook :: ToGenericHook ModifyingHook where
+  toGenericHook (ModifyingHook { shouldProcessNode, processNode }) =
```
Comment (kaol) [inline packages/components/src/HtmlRenderer/Models.purs:105]: Remove trailing white space.

File: packages/components/src/HtmlRenderer/Models.purs [hunk 14]
```diff
@@ -0,0 +1,639 @@
...
+  , tabindex          :: Maybe String
+  , target            :: Maybe String
+  , title             :: Maybe String
+  , type              :: Maybe String
+  , usemap            :: Maybe String
+  , value             :: Maybe String
+  , width             :: Maybe String
+  , wmode             :: Maybe String
+  , wrap              :: Maybe String
+  , onclick           :: Maybe String
```
Comment (kaol) [inline packages/components/src/HtmlRenderer/Models.purs:257]: Remove trailing white space.

File: packages/components/src/Lettera.purs [hunk 15]
```diff
@@ -56,6 +60,20 @@ letteraTagUrl = letteraBaseUrl <> "/list/tag/"
 letteraSearchUrl :: String
 letteraSearchUrl = letteraBaseUrl <> "/list/search"
 
+type LetteraError =
+  { type        :: LetteraErrorType
+  , information :: Maybe String
+  }
```
Comment (kaol) [inline packages/components/src/Lettera.purs:66]: Remove trailing white space.

File: packages/components/src/Lettera.purs [hunk 16]
```diff
@@ -139,6 +157,19 @@ getDraftArticle aptomaId { time, publication, user, hash } = do
         pure $ Left "Unauthorized"
       | (StatusCode s) <- response.status -> pure $ Left $ "Unexpected HTTP status: " <> show s
 
+getFrontpageHtml :: Paper -> String -> Aff (Either LetteraError String)
+getFrontpageHtml paper category = do
+  let request = letteraFrontPageHtmlUrl <> "?paper=" <> Paper.toString paper  <> "&category=" <> category
+  htmlResponse <- AX.get ResponseFormat.string request
+  case htmlResponse of
+    Left err ->
```
Comment (kaol) [inline packages/components/src/Lettera.purs:165]: Remove trailing white space.

File: apps/mosaico/src/Main.purs [hunk 17]
```diff
@@ -307,9 +311,27 @@ frontpage env { guards: { credentials } } = do
             appendMosaico mosaicoString env.htmlTemplate >>= appendHead (mkWindowVariables windowVars)
   pure $ maybeInvalidateAuth user $ htmlContent $ Response.ok $ StringBody html
 
-mkArticleFeed :: Maybe String -> String -> Array ArticleStub -> String
-mkArticleFeed feedPage feedType feedContent =
-  stringify $ encodeJson { feedPage, feedType, feedContent: encodeStringifyArticleStubs feedContent }
+getFrontPage :: Paper -> String -> Aff ArticleFeed
```
Comment (kaol) [inline apps/mosaico/src/Main.purs:314]: We've used frontpage as "Frontpage" in other camel cased contexts, like in "FrontpageFeed". Let's stick to the same convention with the new names that this PR introduces, so this function would be `getFrontpage`. The same goes for other FrontPages.

File: packages/components/src/HtmlRenderer.purs [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+type Self =
+  { props :: Props
+  }
+
+type Props =
+  { content :: String
+  , hooks   :: Maybe (Array Models.HookRep)
+  }
+
+htmlRendererComponent :: Component Props
```
Comment (kaol) [inline packages/components/src/HtmlRenderer.purs:30]: Let's just do this as pure `render` function. A component with no state is really an antipattern with React Hooks, this only adds complexity for no benefit.

**Review by kaol — CHANGES_REQUESTED** [review]

**Review by kaol — APPROVED** [review]

**Pull request merged**